### PR TITLE
Suggest adding super trait constraints

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

**Comment by estebank** [comment]
The first commit is the only logic change. The second change only changes the output to be more verbose and clear.

File: src/librustc_trait_selection/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -173,6 +173,7 @@ fn suggest_restriction(
     fn_sig: Option<&hir::FnSig<'_>>,
     projection: Option<&ty::ProjectionTy<'_>>,
     trait_ref: ty::PolyTraitRef<'_>,
+    super_traits: Option<(&Ident, &hir::GenericBounds<'_>)>,
```
Comment (nikomatsakis) [inline src/librustc_trait_selection/traits/error_reporting/suggestions.rs:176]: It would be helpful to have a comment with what `super_traits` represents.

Comment (nikomatsakis) [reply src/librustc_trait_selection/traits/error_reporting/suggestions.rs:176]: Ideally a code example showing what it is.

File: src/librustc_trait_selection/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -288,13 +307,33 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         let mut hir_id = body_id;
         while let Some(node) = self.tcx.hir().find(hir_id) {
             match node {
+                hir::Node::Item(hir::Item {
+                    ident,
+                    kind: hir::ItemKind::Trait(_, _, generics, bounds, _),
+                    ..
+                }) if param_ty && self_ty == self.tcx.types.self_param => {
```
Comment (nikomatsakis) [inline src/librustc_trait_selection/traits/error_reporting/suggestions.rs:314]: I don't think you need the `param_ty &&` part of this test, it's redundant, isn't it? Maybe move it to an `assert!`?

File: src/librustc_trait_selection/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -288,13 +307,33 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+                        Some((ident, bounds)),
+                    );
+                    return;
+                }
+
                 hir::Node::TraitItem(hir::TraitItem {
                     generics,
                     kind: hir::TraitItemKind::Fn(..),
                     ..
                 }) if param_ty && self_ty == self.tcx.types.self_param => {
```
Comment (nikomatsakis) [inline src/librustc_trait_selection/traits/error_reporting/suggestions.rs:332]: (same here...)

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me with nits

**Pull request merged**